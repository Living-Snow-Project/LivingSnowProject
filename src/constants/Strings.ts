// for TextInput placeholder prop
const Placeholders = {
  GPS: {
    AcquiringLocation: "Looking for GPS signal",
    EnterCoordinates: "ie. 12.345678, -123.456789",
    NoPermissions:
      "Check location permissions in Settings. Enter coordinates manually.",
    NoLocation: "Could not determine location. Enter coordinates manually.",
  },
  // TODO: RecordScreen placeholders
};

const Labels = {
  TimelineScreen: {
    DownloadedRecords: "Downloaded",
    NoRecords: "No records to display",
    PendingRecords: "Pending",
  },
  StatusBar: {
    NoConnection: "No Internet Connection",
  },
};

// for Alert.alert calls
const Notifications = {
  uploadSuccess: {
    title: "Upload succeeded",
    message: "Thanks for your submission.",
  },
  uploadFailed: {
    title: "Record Saved",
    message: "We will upload it later.",
  },
  invalidCoordinates: {
    title: "Invalid GPS coordinates",
    message:
      'Coordinates must be in "lat, long" format. ie. 12.345678, -123.456789',
  },
};

export { Notifications, Placeholders, Labels };
