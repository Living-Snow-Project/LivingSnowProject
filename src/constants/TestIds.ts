const TestIds = {
  GPS: {
    gpsManualInputTestId: "gps-manual-input",
    gpsManualPressableTestId: "gps-manual-pressable",
  },
  Photos: {
    photoSelectorTestId: "photo-selector",
  },
  Pickers: {
    recordSelectorTestId: "record-type-picker",
    atlastSelectorTestId: "atlas-type-picker",
  },
  RecordScreen: {
    UploadButton: "record-screen-upload-button",
  },
  TimelineScreen: {
    RefreshControl: "timeline-screen-refresh-control",
    SettingsButton: "timeline-screen-settings-button",
    NewRecordButton: "timeline-screen-record-button",
  },
  SettingsScreen: {
    ShowGpsWarning: "settings-screen-show-gps-warning",
    ShowAtlasRecords: "settings-screen-show-atlas-records",
    ShowOnlyAtlasRecords: "settings-screen-show-only-atlas-records",
  },
  Icons: {
    SampleIcon: "sample-icon",
    SightingIcon: "sighting-icon",
    PictureIcon: "picture-icon",
    DeleteIcon: "delete-icon",
    EditIcon: "edit-icon",
    SnowIcon: "snow-icon",
  },
  RecordList: {
    EditRecordAction: "edit-record-action",
    DeleteRecordAction: "delete-record-action",
  },
};

export default TestIds;
