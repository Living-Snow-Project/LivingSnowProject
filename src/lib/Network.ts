import Logger from "./Logger";
import serviceEndpoint from "../constants/Service";
import { Record } from "../record/Record";

const recordsUri: string = `${serviceEndpoint}/api/records/`;
const uploadPhotoUri = (id: number): string => `${recordsUri}${id}/photo`;
const downloadPhotoUri = (id: number): string =>
  `${serviceEndpoint}/api/photos/${id}`;

function dumpRecord(record: Record): void {
  Logger.Info(
    `Handling POST Request: ${serviceEndpoint}/api/records` +
      `\n  Type: ${record.type}` +
      `\n  Name: ${record.name}` +
      `\n  Date: ${record.date}` +
      `\n  Org: ${record.organization}` +
      `\n  Atlas Type: ${record.atlasType}` +
      `\n  TubeId: ${record.tubeId}` +
      `\n  Latitude: ${record.latitude}` +
      `\n  Longitude: ${record.longitude}` +
      `\n  Description: ${record.locationDescription}` +
      `\n  Notes: ${record.notes}` +
      `\n JSON Body:\n${JSON.stringify(record)}`
  );
}

function failedFetch(operation: string, response: Response): Promise<void> {
  const messagePrefix = `Failed to ${operation}.`;

  if (response?.status) {
    const error: string = `${messagePrefix} ${response.status}: ${response.statusText}`;
    Logger.Error(error);
    return Promise.reject(error);
  }

  Logger.Error(`${messagePrefix}: ${response}`);
  return Promise.reject(response);
}

async function uploadRecord(record: Record): Promise<Record> {
  const operation = `uploadRecord`;
  dumpRecord(record);

  return fetch(recordsUri, {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(record),
  })
    .then((response) =>
      response.ok ? response.json() : failedFetch(operation, response)
    )
    .catch((error) => failedFetch(operation, error));
}

async function uploadPhoto({ id, photoStream }: Photo): Promise<void> {
  const operation = `uploadPhoto`;

  return fetch(uploadPhotoUri(id), {
    method: "POST",
    headers: {
      "Content-Type": "image/jpeg",
    },
    body: photoStream,
  })
    .then((response) =>
      response.ok ? Promise.resolve() : failedFetch(operation, response)
    )
    .catch((error) => failedFetch(operation, error));
}

// TODO: replace any with Record[] when type alignment happens
async function downloadRecords(): Promise<any> {
  const operation = `downloadRecords`;

  Logger.Info(`Handling GET Request: ${recordsUri}`);

  return fetch(recordsUri)
    .then((response) =>
      response.ok ? response.json() : failedFetch(operation, response)
    )
    .catch((error) => failedFetch(operation, error));
}

export { downloadPhotoUri, uploadRecord, uploadPhoto, downloadRecords };
