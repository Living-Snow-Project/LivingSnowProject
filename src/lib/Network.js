import { serviceEndpoint } from "../constants/Service";

const recordsUri = `${serviceEndpoint}/api/records/`;
const uploadPhotoUri = (id) => `${recordsUri}${id}/photo`;
export const downloadPhotoUri = (id) => `${serviceEndpoint}/api/photos/${id}`;

function dumpRecord(record) {
  console.log(
    `Handling POST Request: ${serviceEndpoint}/api/records` +
      `\n  Type: ${record.type}` +
      `\n  Name: ${record.name}` +
      `\n  Date: ${record.date}` +
      `\n  Org: ${record.organization}` +
      `\n  Atlas Type: ${record.atlasType}` +
      `\n  TubeId: ${record.tubeId}` +
      `\n  Latitude: ${record.latitude}` +
      `\n  Longitude: ${record.longitude}` +
      `\n  Description: ${record.locationDescription}` +
      `\n  Notes: ${record.notes}` +
      `\nJSON Body\n  `,
    record
  );
}

function failedFetch(operation, response) {
  const messagePrefix = `Failed to ${operation}.`;

  if (response?.status) {
    const error = `${messagePrefix} ${response.status}: ${response.statusText}`;
    console.log(error);
    return Promise.reject(error);
  }

  console.log(messagePrefix, response);
  return Promise.reject(response);
}

export class Network {
  // returns a resolved Promise<record> on success
  static async uploadRecord(record) {
    const operation = `uploadRecord`;
    dumpRecord(record);

    return fetch(recordsUri, {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(record),
    })
      .then((response) =>
        response.ok ? response.json() : failedFetch(operation, response)
      )
      .catch((error) => failedFetch(operation, error));
  }

  // photoStream = {uri, width, height}
  // returns a resolved Promise<void> on success
  static async uploadPhoto({ id, photoStream }) {
    const operation = `uploadPhoto`;

    return fetch(uploadPhotoUri(id), {
      method: "POST",
      headers: {
        "Content-Type": "image/jpeg",
      },
      body: photoStream,
    })
      .then((response) =>
        response.ok ? Promise.resolve() : failedFetch(operation, response)
      )
      .catch((error) => failedFetch(operation, error));
  }

  // returns a resolved Promise<records> on success
  static async downloadRecords() {
    const operation = `downloadRecords`;

    console.log(`Handling GET Request: ${recordsUri}`);

    return fetch(recordsUri)
      .then((response) =>
        response.ok ? response.json() : failedFetch(operation, response)
      )
      .catch((error) => failedFetch(operation, error));
  }
}
