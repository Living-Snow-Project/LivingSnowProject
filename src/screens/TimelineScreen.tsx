import React, { useContext, useEffect, useRef, useState } from "react";
import { FlatList, View } from "react-native";
import NetInfo from "@react-native-community/netinfo";
import Logger from "../lib/Logger";
import StatusBar from "../components/StatusBar";
import { ExampleRecordList } from "../components/RecordList";
import styles from "../styles/Timeline";
import TestIds from "../constants/TestIds";
import {
  RecordReducerActionsContext,
  RecordReducerStateContext,
} from "../hooks/useRecordReducer";
import useRecordList from "../hooks/useRecordList";
import PressableOpacity from "../components/PressableOpacity";
import { ScrollTopIcon } from "../components/Icons";

function Separator() {
  return <View style={styles.separator} />;
}

type ContainerDimensions = {
  width: number;
  height: number;
};

export default function TimelineScreen({ navigation }) {
  const scrollViewRef = useRef<FlatList>(null);
  const [containerDims, setContainerDims] = useState<ContainerDimensions>({
    width: 0,
    height: 0,
  });
  const [scrollingToTop, setScrollingToTop] = useState<boolean>(false);
  const [showScrollToTop, setShowScrollToTop] = useState<boolean>(false);

  const recordList = useRecordList(navigation);
  const [connected, setConnected] = useState<boolean>(true);
  const [refreshing, setRefreshing] = useState<boolean>(false);
  const recordReducerActionsContext = useContext(RecordReducerActionsContext);
  const recordReducerStateContext = useContext(RecordReducerStateContext);

  // TODO: should be in App component and in a Reducer\Context
  useEffect(
    () =>
      NetInfo.addEventListener(({ isConnected }) => {
        setConnected(!!isConnected);

        if (isConnected) {
          setRefreshing(true);
        }
      }),
    []
  );

  useEffect(() => {
    if (!refreshing) {
      return;
    }

    if (!connected) {
      setRefreshing(false);
      return;
    }

    recordReducerActionsContext
      .retryPendingRecords()
      .then(() => recordReducerActionsContext.downloadRecords())
      .catch(() =>
        Logger.Warn(`Could not download records. Please try again later.`)
      )
      .finally(() => setRefreshing(false));
  }, [refreshing]);

  return (
    <>
      <View
        style={styles.container}
        onLayout={({ nativeEvent }) => setContainerDims(nativeEvent.layout)}
      >
        <StatusBar
          state={recordReducerStateContext.state}
          isConnected={connected}
        />
        <FlatList
          ref={scrollViewRef}
          data={recordList}
          testID={TestIds.TimelineScreen.FlatList}
          renderItem={({ item }) => item}
          keyExtractor={(item, index) => `${index}`}
          ListEmptyComponent={ExampleRecordList}
          ItemSeparatorComponent={Separator}
          onScroll={({ nativeEvent }) => {
            if (nativeEvent.contentOffset.y < 1000) {
              setScrollingToTop(false);
            }
            setShowScrollToTop(nativeEvent.contentOffset.y > 1000);
          }}
          onRefresh={() => setRefreshing(true)}
          refreshing={false} // StatusBar component handles activity indicator
          onEndReached={() => {
            // keep an eye on this (if list is "empty" and it gets called)
            const { downloadedRecords } = recordReducerStateContext;
            recordReducerActionsContext.downloadNextRecords(
              downloadedRecords[downloadedRecords.length - 1].date
            );
          }}
          onEndReachedThreshold={0.5}
        />
      </View>
      {showScrollToTop && !scrollingToTop && (
        <View
          style={[
            styles.scrollToTop,
            {
              top: containerDims.height - 65,
              left: containerDims.width / 2 - 25,
            },
          ]}
        >
          <PressableOpacity
            testID={TestIds.TimelineScreen.ScrollToTopButton}
            onPress={() => {
              setScrollingToTop(true);
              scrollViewRef?.current?.scrollToOffset({
                animated: true,
                offset: 0,
              });
            }}
          >
            <ScrollTopIcon />
          </PressableOpacity>
        </View>
      )}
    </>
  );
}
