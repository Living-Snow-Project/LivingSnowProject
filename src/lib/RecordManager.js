import { Network } from './Network';
import { Storage } from './Storage';

export class RecordManager {
  static async uploadRecord(record, photoUris) {
    return Network.uploadRecord(record)
    .then(response => {
      return photoUris.reduce((promise, uri) => {
        return promise.then(() => {
          let photo = {id: response.id, photoStream: uri};
          return Network.uploadPhoto(photo).catch(async error => {
            console.log(error);
            await Storage.savePhoto(photo);
          })
        })
      }, Promise.resolve()).then(() => Promise.resolve());
    })
    .catch(async error => {
      console.log(error);
      record.photoUris = photoUris;
      await Storage.saveRecord(record);
      return Promise.reject(error);
    });
  }

  // uploads records that were saved while user was offline
  static async retryRecords() {
    return Storage.loadRecords().then(async records => {
      if (!records) {
        return;
      }

      await Storage.clearRecords();
      
      records.reduce((promise, record) => {
        return promise.then(() => {
          let photoUris = record?.photoUris;
          record.photoUris = null;

          return RecordManager.uploadRecord(record, photoUris)
          .catch(error => console.log(`continue records reducer to prevent data loss: ${error}`));
        });
      }, Promise.resolve())
      .catch(error => console.log(`error in record reducer: ${error}`));
    });
  }

  // uploads photos that were saved while user was offline
  static async retryPhotos() {
    return Storage.loadPhotos().then(async photos => {
      if (!photos) {
        return;
      }

      await Storage.clearPhotos();

      photos.reduce((promise, photo) => {
        return promise.then(() => {
          return Network.uploadPhoto(photo).catch(error => {
            console.log(error);
            Storage.savePhoto(photo);
          });
        });
      }, Promise.resolve())
      .catch(error => console.log(`error in photo reducer: ${error}`));
    });
  }
}
