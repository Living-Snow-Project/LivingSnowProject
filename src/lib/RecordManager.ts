import * as Network from "./Network";
import {
  loadRecords,
  saveRecord,
  clearRecords,
  loadPhotos,
  savePhoto,
  clearPhotos,
} from "./Storage";
import { jsonToRecord } from "../record/Record";
import Logger from "./Logger";

// returns the AlgaeRecord server responds with
// rejects with an error string
async function uploadRecord(
  record: AlgaeRecord,
  photos: Photo[]
): Promise<AlgaeRecord> {
  let uploadPhotoError;
  const localRecord = jsonToRecord<AlgaeRecord>(JSON.stringify(record));
  localRecord.id = 0;
  return Network.uploadRecord(localRecord)
    .then((response) =>
      photos
        .reduce(
          (promise, photo) =>
            promise.then(() => {
              // picks up the record id of photo associated with
              const pendingPhoto: PendingPhoto = { id: response.id, ...photo };
              return Network.uploadPhoto(pendingPhoto).catch(async (error) => {
                Logger.Warn(`${error}`);
                uploadPhotoError = error;
                await savePhoto(pendingPhoto);
              });
            }),
          Promise.resolve()
        )
        .then(() => Promise.resolve(response))
    )
    .catch(async (error) => {
      Logger.Warn(`${error}`);
      localRecord.id = record.id;
      localRecord.photos = photos;
      await saveRecord(localRecord);
      return Promise.reject(error);
    })
    .finally(() =>
      !uploadPhotoError ? Promise.resolve() : Promise.reject(uploadPhotoError)
    );
}

// uploads photos that were saved while user was offline
async function retryPhotos(): Promise<void> {
  return loadPhotos().then(async (photos) => {
    if (photos.length === 0) {
      return Promise.resolve();
    }

    await clearPhotos();

    return photos.reduce(
      (promise, photo) =>
        promise
          .then(() => Network.uploadPhoto(photo))
          .catch(async (error) => {
            Logger.Warn(
              `Network.uploadPhoto rejected: continue photo reducer to prevent data loss: ${error}`
            );
            await savePhoto(photo);
          }),
      Promise.resolve()
    );
  });
}

// uploads records that were saved while user was offline
// returns the current array of pending records
async function retryPendingRecords(): Promise<AlgaeRecord[]> {
  return loadRecords()
    .then(async (records) => {
      if (records.length === 0) {
        return Promise.resolve([]);
      }

      await clearRecords();

      return records.reduce(
        (promise, record) =>
          promise
            .then(() => {
              const { photos } = record;
              // @ts-ignore (currently photos will always be [] and never undefined)
              return uploadRecord(record, photos);
            })
            .catch((error) =>
              Logger.Warn(
                `uploadRecord rejected: continue records reducer to prevent data loss: ${error}`
              )
            ),
        Promise.resolve()
      );
    })
    .then(() => retryPhotos().then(() => loadRecords()));
}

export { uploadRecord, retryPendingRecords };
