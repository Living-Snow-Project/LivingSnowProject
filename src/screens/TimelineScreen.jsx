import React, { useCallback, useEffect, useState } from "react";
import {
  RefreshControl,
  ScrollView,
  StyleSheet,
  Text,
  View,
} from "react-native";
import { useFocusEffect } from "@react-navigation/native";
import NetInfo from "@react-native-community/netinfo";
import Storage from "../lib/Storage";
import { Network } from "../lib/Network";
import RecordManager from "../lib/RecordManager";
import TimelineRow from "../components/TimelineRow";
import StatusBar from "../components/StatusBar";

export default function TimelineScreen({ navigation }) {
  const [connected, setConnected] = useState(true);
  const [refreshing, setRefreshing] = useState(false);
  const [pendingRecords, setPendingRecords] = useState([]);
  const [downloadedRecords, setDownloadedRecords] = useState([]);
  const [status, setStatus] = useState({
    text: null,
    type: null,
    onDone: null,
  });

  useEffect(() => {
    const unsubscribe = NetInfo.addEventListener(({ isConnected }) => {
      setConnected(isConnected);

      if (isConnected) {
        clearStatus();
        setRefreshing(true);
      } else {
        updateStatus(`No Internet Connection`, `static`, null);
      }
    });

    return unsubscribe;
  }, []);

  const displaySavedRecords = useCallback(() => {
    // records.photoUris is saved on disk as array of {uri, width, height} but RecordDetailsScreen expects `uri;uri;...` string format
    Storage.loadRecords().then((records) => {
      records.forEach((record, index) => {
        records[index].photoUris = record.photoUris.reduce(
          (accumulator, photo) => `${accumulator}${photo.uri};`,
          ``
        );
      });
      setPendingRecords(records);
    });
  }, []);

  // force a re-render when user comes back from Settings screen
  // TODO: figure out why this is getting called on blur too
  useFocusEffect(displaySavedRecords);

  useEffect(() => {
    if (!refreshing) {
      return;
    }

    if (!connected) {
      setRefreshing(false);
      return;
    }

    RecordManager.retryRecords()
      .then(() => RecordManager.retryPhotos())
      .then(() => Network.downloadRecords())
      .then((response) => setDownloadedRecords(response))
      .catch(() =>
        console.log(`Could not download records. Please try again later.`)
      )
      .finally(() => {
        setRefreshing(false);
        displaySavedRecords();
      });
  }, [refreshing]);

  const renderRecords = useCallback((records, label) => {
    if (records.length === 0) {
      return null;
    }

    return (
      <View>
        <View style={styles.recordStatusContainer}>
          <Text style={styles.recordStatusText}>{label}</Text>
        </View>
        {records.map((record, index) => (
          <TimelineRow
            key={index}
            navigation={navigation}
            record={record}
            showAll={label.includes(`Pending`)}
          />
        ))}
      </View>
    );
  }, []);

  const updateStatus = (text, type, onDone = null) =>
    setStatus({ text, type, onDone });
  const clearStatus = () => updateStatus(null, null, null);

  return (
    <View style={styles.container}>
      <StatusBar text={status.text} type={status.type} onDone={status.onDone} />
      <ScrollView
        style={styles.container}
        refreshControl={
          <RefreshControl
            refreshing={refreshing}
            onRefresh={() => setRefreshing(true)}
          />
        }
      >
        {pendingRecords.length === 0 && downloadedRecords.length === 0 && (
          <Text style={styles.noRecords}>No records to display</Text>
        )}
        {renderRecords(pendingRecords, `Pending`)}
        {renderRecords(downloadedRecords, `Downloaded`)}
      </ScrollView>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
  },
  noRecords: {
    textAlign: "center",
    marginTop: 20,
  },
  recordStatusContainer: {
    backgroundColor: "lightgrey",
    borderBottomWidth: 1,
  },
  recordStatusText: {
    fontSize: 16,
    textAlign: "center",
  },
});
