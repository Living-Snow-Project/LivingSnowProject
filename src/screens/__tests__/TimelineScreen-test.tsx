import React from "react";
import { act, fireEvent, render, waitFor } from "@testing-library/react-native";
import NetInfo, {
  NetInfoChangeHandler,
  NetInfoState,
} from "@react-native-community/netinfo";
import TimelineScreen from "../TimelineScreen";
import {
  SettingsButton,
  NewRecordButton,
} from "../../navigation/MainTabNavigator";
import * as RecordManager from "../../lib/RecordManager";
import { makeExampleRecord } from "../../record/Record";
import * as Storage from "../../lib/Storage";
import * as Network from "../../lib/Network";
import TestIds from "../../constants/TestIds";
import { Labels } from "../../constants/Strings";
import { getAppSettings, setAppSettings } from "../../../AppSettings";
import { mockedNavigate } from "../../../jesttest.setup";

// TimelineScreen takes navigation input prop
const navigation = {
  navigate: jest.fn(),
  addListener: () => () => {},
};

const downloadedRecord: AlgaeRecord = {
  ...makeExampleRecord("Sample"),
  id: 1,
};

const downloadedAtlasRecord: AlgaeRecord = {
  ...makeExampleRecord("Atlas: Red Dot"),
  id: 2,
};

const pendingRecord: AlgaeRecord = {
  ...makeExampleRecord("Sighting"),
  id: 3,
};

const pendingAtlasRecord: AlgaeRecord = {
  ...makeExampleRecord("Atlas: Blue Dot with Sample"),
  id: 4,
};

const setIsConntected = (isConnected: boolean): void => {
  jest.spyOn(NetInfo, "addEventListener").mockImplementationOnce((callback) => {
    callback({ isConnected } as NetInfoState);

    return () => {};
  });
};

const setupDownloadSuccess = () => {
  const retryRecordsSpy = jest
    .spyOn(RecordManager, "retryRecords")
    .mockImplementationOnce(() => Promise.resolve());

  const downloadRecordsSpy = jest
    .spyOn(Network, "downloadRecords")
    .mockImplementationOnce(() =>
      Promise.resolve([downloadedRecord, downloadedAtlasRecord])
    );

  setIsConntected(true);

  return {
    retryRecordsSpy,
    downloadRecordsSpy,
  };
};

const setupDownloadFailed = () => {
  const retryRecordsSpy = jest
    .spyOn(RecordManager, "retryRecords")
    .mockImplementationOnce(() => Promise.reject());

  setIsConntected(true);

  return retryRecordsSpy;
};

describe("TimelineScreen test suite", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test("renders", () => {
    const { toJSON } = render(<TimelineScreen navigation={navigation} />);
    expect(toJSON()).toMatchSnapshot();
  });

  test("download records succeeds", async () => {
    const { retryRecordsSpy, downloadRecordsSpy } = setupDownloadSuccess();

    const { getByTestId, getByText } = render(
      <TimelineScreen navigation={navigation} />
    );

    await waitFor(() => getByTestId(downloadedRecord.id.toString()));

    expect(retryRecordsSpy).toBeCalledTimes(1);
    expect(downloadRecordsSpy).toBeCalledTimes(1);
    expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();
  });

  test("download records fails", async () => {
    const retryRecordsSpy = setupDownloadFailed();

    const { getByText } = render(<TimelineScreen navigation={navigation} />);

    await waitFor(() => getByText(Labels.TimelineScreen.ExampleRecords));
    expect(retryRecordsSpy).toBeCalledTimes(1);
  });

  test("pending records", async () => {
    const retryRecordsSpy = setupDownloadFailed();
    jest
      .spyOn(Storage, "loadRecords")
      .mockImplementationOnce(() =>
        Promise.resolve([pendingRecord, pendingAtlasRecord])
      );

    const { getByTestId, getByText } = render(
      <TimelineScreen navigation={navigation} />
    );

    // non-atlas visible
    await waitFor(() => getByTestId(pendingRecord.id.toString()));
    expect(retryRecordsSpy).toBeCalledTimes(1);
    expect(getByText(Labels.TimelineScreen.PendingRecords)).toBeTruthy();

    // atlas visible
    expect(getByTestId(pendingAtlasRecord.id.toString())).toBeTruthy();
  });

  describe("atlas scenarios", () => {
    let prevAppSettings: AppSettings;

    beforeEach(() => {
      prevAppSettings = getAppSettings();
    });

    afterEach(() => setAppSettings(prevAppSettings));

    test("non-atlas and atlas records are displayed", async () => {
      const { retryRecordsSpy, downloadRecordsSpy } = setupDownloadSuccess();

      setAppSettings((prev) => ({
        ...prev,
        showAtlasRecords: true,
        showOnlyAtlasRecords: false,
      }));

      const { getByTestId, getByText } = render(
        <TimelineScreen navigation={navigation} />
      );

      // non-atlas visible
      await waitFor(() => getByTestId(downloadedRecord.id.toString()));

      expect(retryRecordsSpy).toBeCalledTimes(1);
      expect(downloadRecordsSpy).toBeCalledTimes(1);
      expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();

      // atlas visible
      expect(getByTestId(downloadedAtlasRecord.id.toString())).toBeTruthy();
    });

    test("atlas records are not displayed", async () => {
      const { retryRecordsSpy, downloadRecordsSpy } = setupDownloadSuccess();

      setAppSettings((prev) => ({
        ...prev,
        showAtlasRecords: false,
        showOnlyAtlasRecords: false,
      }));

      const { getByTestId, getByText, queryByTestId } = render(
        <TimelineScreen navigation={navigation} />
      );

      // non-atlas visible
      await waitFor(() => getByTestId(downloadedRecord.id.toString()));

      expect(retryRecordsSpy).toBeCalledTimes(1);
      expect(downloadRecordsSpy).toBeCalledTimes(1);
      expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();

      // atlas not visible
      expect(queryByTestId(downloadedAtlasRecord.id.toString())).toBeFalsy();
    });

    test("only atlas records are displayed", async () => {
      const { retryRecordsSpy, downloadRecordsSpy } = setupDownloadSuccess();

      setAppSettings((prev) => ({
        ...prev,
        showAtlasRecords: true,
        showOnlyAtlasRecords: true,
      }));

      const { getByTestId, getByText, queryByTestId } = render(
        <TimelineScreen navigation={navigation} />
      );

      // atlas visible
      await waitFor(() => getByTestId(downloadedAtlasRecord.id.toString()));

      expect(retryRecordsSpy).toBeCalledTimes(1);
      expect(downloadRecordsSpy).toBeCalledTimes(1);
      expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();

      // non-atlas not visible
      expect(queryByTestId(downloadedRecord.id.toString())).toBeFalsy();
    });
  });

  test("connection lost", async () => {
    setIsConntected(false);
    const { getByText } = render(<TimelineScreen navigation={navigation} />);

    await waitFor(() => getByText(Labels.StatusBar.NoConnection));
    expect(getByText(Labels.StatusBar.NoConnection)).toBeTruthy();
  });

  test("connection restored", async () => {
    let cacheCallback: NetInfoChangeHandler;

    jest
      .spyOn(NetInfo, "addEventListener")
      .mockImplementationOnce((callback) => {
        cacheCallback = callback;
        callback({ isConnected: false } as NetInfoState);
        return () => {};
      });

    const { getByText, queryByText } = render(
      <TimelineScreen navigation={navigation} />
    );

    await waitFor(() => getByText(Labels.StatusBar.NoConnection));
    expect(getByText(Labels.StatusBar.NoConnection)).toBeTruthy();

    await act(async () => cacheCallback({ isConnected: true } as NetInfoState));
    expect(queryByText(Labels.StatusBar.NoConnection)).toBeFalsy();
  });

  test("pull to refresh with connection", async () => {
    setupDownloadSuccess();

    const { getByText, getByTestId } = render(
      <TimelineScreen navigation={navigation} />
    );

    await waitFor(() => getByTestId(downloadedRecord.id.toString()));
    expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();

    await act(async () =>
      getByTestId(
        TestIds.TimelineScreen.RefreshControl
      ).props.refreshControl.props.onRefresh()
    );

    expect(getByText(Labels.TimelineScreen.DownloadedRecords)).toBeTruthy();
  });

  test("pull to refresh without connection", async () => {
    setIsConntected(false);

    const { getByTestId } = render(<TimelineScreen navigation={navigation} />);

    // GO GO RNTL! "don't test the implementation..."
    await act(async () =>
      getByTestId(
        TestIds.TimelineScreen.RefreshControl
      ).props.refreshControl.props.onRefresh()
    );

    // nothing to assert, strictly for code coverage
    // TODO: re-evaluate setPendingRecords initialization and displaySavedRecords()
  });

  test("navigate back to TimelineScreen", async () => {
    let cb: () => void;
    const localNavigation = {
      navigate: jest.fn(),
      addListener: (event: string, callback: () => void) => {
        expect(event).toEqual("focus");
        cb = callback;
      },
    };

    render(<TimelineScreen navigation={localNavigation} />);
    await waitFor(() => expect(cb).not.toBeUndefined());
    await act(async () => cb());
  });

  test("navigate to SettingsScreen", () => {
    const navigate = jest.fn();
    const { getByTestId } = render(<SettingsButton navigate={navigate} />);
    fireEvent.press(getByTestId(TestIds.TimelineScreen.SettingsButton));
    expect(navigate).toBeCalledWith("Settings");
  });

  test("navigate to RecordScreen", async () => {
    const navigate = jest.fn();
    const { getByTestId } = render(<NewRecordButton navigate={navigate} />);
    fireEvent.press(getByTestId(TestIds.TimelineScreen.NewRecordButton));
    expect(navigate).toBeCalledWith("Record");
  });

  test("navigate to record details screen", async () => {
    const { retryRecordsSpy, downloadRecordsSpy } = setupDownloadSuccess();

    const { getByTestId } = render(<TimelineScreen navigation={navigation} />);

    await waitFor(() => getByTestId(downloadedRecord.id.toString()));

    fireEvent.press(getByTestId(downloadedRecord.id.toString()));

    expect(retryRecordsSpy).toBeCalledTimes(1);
    expect(downloadRecordsSpy).toBeCalledTimes(1);
    expect(mockedNavigate).toBeCalledWith("RecordDetails", {
      record: downloadedRecord,
    });
  });
});
