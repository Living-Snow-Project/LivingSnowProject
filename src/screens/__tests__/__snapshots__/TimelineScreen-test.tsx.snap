// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineScreen tests renders 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
      }
    }
  >
    <RCTRefreshControl />
    <View>
      <Text
        style={
          Object {
            "marginTop": 20,
            "textAlign": "center",
          }
        }
      >
        No records to display
      </Text>
    </View>
  </RCTScrollView>
</View>
`;
